pkg/cvo/availableupdates_test.go

```diff
@@ -0,0 +1,306 @@
+	panic("implement me")
+}
+
+func (q queueStub) ShutDown() {
+	panic("implement me")
+}
+
+func (q queueStub) ShutDownWithDrain() {
+	panic("implement me")
+}
+
+func (q queueStub) ShuttingDown() bool {
+	panic("implement me")
+}
+
+func (q queueStub) AddAfter(interface{}, time.Duration) {
+	panic("implement me")
+}
+
+func (q queueStub) AddRateLimited(interface{}) {
+	panic("implement me")
+}
+
+func (q queueStub) Forget(interface{}) {
+	panic("implement me")
+}
+
+func (q queueStub) NumRequeues(interface{}) int {
+	panic("implement me")
+}
+
+// notFoundProxyLister is a stub for ProxyLister
+type notFoundProxyLister struct{}
+
+func (n notFoundProxyLister) List(labels.Selector) ([]*configv1.Proxy, error) {
+	return nil, nil
+}
+
+func (n notFoundProxyLister) Get(name string) (*configv1.Proxy, error) {
+	return nil, errors.NewNotFound(schema.GroupResource{Group: configv1.GroupName, Resource: "proxy"}, name)
+}
+
+type notFoundConfigMapLister struct{}
+
+func (n notFoundConfigMapLister) List(labels.Selector) ([]*corev1.ConfigMap, error) {
+	return nil, nil
+}
+
+func (n notFoundConfigMapLister) Get(name string) (*corev1.ConfigMap, error) {
+	return nil, errors.NewNotFound(schema.GroupResource{Group: "", Resource: "configmap"}, name)
+}
+
+type fakeConditionRegistry struct{}
+
+func (f fakeConditionRegistry) Register(string, clusterconditions.Condition) {
+	panic("implement me")
+}
+
+func (f fakeConditionRegistry) PruneInvalid(ctx context.Context, matchingRules []configv1.ClusterCondition) ([]configv1.ClusterCondition, error) {
+	return matchingRules, nil
+}
+
+type clusterConditionRuleType string
+
+const (
+	ruleTypeAlways clusterConditionRuleType = "Always"
+	ruleTypePromQL clusterConditionRuleType = "PromQL"
+)
+
+type clusterConditionRuleFakePromql string
+
+const (
+	evalToYes   clusterConditionRuleFakePromql = "YES"
+	evalToNo    clusterConditionRuleFakePromql = "NO"
+	errorOnEval clusterConditionRuleFakePromql = "ERROR"
+	skip        clusterConditionRuleFakePromql = "SKIP"
+)
+
+func (f fakeConditionRegistry) Match(ctx context.Context, matchingRules []configv1.ClusterCondition) (bool, error) {
+	for _, rule := range matchingRules {
+		switch clusterConditionRuleType(rule.Type) {
+		case ruleTypeAlways:
+			return true, nil
+		case ruleTypePromQL:
+			switch clusterConditionRuleFakePromql(rule.PromQL.PromQL) {
+			case evalToYes:
+				return true, nil
+			case evalToNo:
+				return false, nil
+			case errorOnEval:
+				return false, fmt.Errorf("ERROR")
+			case skip:
+				continue
+			default:
+				panic("This fake works only with YES, NO, ERROR, and SKIP")
+			}
+		default:
+			panic("This fake works only with Always and PromQL risks")
+		}
+	}
+	return false, nil
+}
+
+func osusWithSingleConditionalEdge(from, to string, ruleType clusterConditionRuleType, promql clusterConditionRuleFakePromql) ([]configv1.ConditionalUpdate, *httptest.Server) {
+	osus := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
nit: we could possibly simplify this test harness a bit if we pull [this][1] out into a new function, because the upstream defaulting, available-updates throttling, and proxy transport lookup don't seem as interesting to cover in tests.  You could create a dummy client to pass through to `calculateAvailableUpdatesStatus` like:

```go
type MockRoundTripper func(r *http.Request) *http.Response

func (f MockRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
    return f(r), nil
}
```

and:

```go
httpClient := &http.Client{
  Transport: MockRoundTripper(func(r *http.Request) *http.Response {
    return &http.Response{
      StatusCode: http.StatusOK,
      Body:       io.NopCloser(strings.NewReader(fmt.Sprintf(`{
  "nodes": [{"version": "%s", "payload": "payload/%s"}, {"version": "%s", "payload": "payload/%s"}],
  "conditionalEdges": [
    {
      "edges": [{"from": "%s", "to": "%s"}],
      "risks": [
        {
          "url": "https://example.com/%s",
          "name": "FourFiveSix",
          "message": "Four Five Five is just fine",
          "matchingRules": [{"type": "%s", "promql": { "promql": "%s"}}]
        }
      ]
    }
  ]
}
`, from, from, to, to, from, to, to, ruleType, promql
      ))),
    },
  },
}
```

or some such without needing to set up an HTTP server that needs to be closed later.

I'm not sure if it's worth the effort to pivot, now that you've already figured out all the plumbing to get this up to the existing level, but it might be worth poking at to see if the pivoting-this-pull cost seems like it might be worth reducing the onboarding-the-next-dev-who-needs-to-understand-this-test-suite cost.

[1]: https://github.com/openshift/cluster-version-operator/blob/53a5d57c2bd3c92c145dac005fab6e4cdcf8f43b/pkg/cvo/availableupdates.go#L71-L90

Reply:
I have considered doing something like that, but the tested functionality felt like deserving a higher-level integration test so I thought test HTTP servers are cheap in Go with `httptest`. The real complexity is in preparing all the data and expected fixtures.

But I'll reconsider, with further development and feedback the code now is different than when started. Thanks for the suggestion!

Reply:
I gave the pivot a quick shot and I think we're not really saving that much complexity, the annoying bits with "set up the input/output" would need to stay, we'd just need different plumbing.

But I actually discovered that I like using `httptest` here. It allows testing the full method and it quite explicitly communicates that the important input comes from a server. The plumbing needed is basically identical to mocking it in the client transport. I liked that I could get rid of the annoying queue stub though :)